src/otx/algo/classification/torchvision_model.py

```diff
@@ -155,45 +154,148 @@ def __init__(
+
+        return head
+
     def forward(
         self,
-        images: torch.Tensor,
-        labels: torch.Tensor | None = None,
+        images: torch.Tensor | dict[str, torch.Tensor],
+        labels: torch.Tensor | dict[str, torch.Tensor] | None = None,
         mode: str = "tensor",
-    ) -> torch.Tensor:
+        **kwargs,
+    ) -> dict[str, tuple | torch.Tensor] | tuple | torch.Tensor:
         """Performs forward pass of the model.
 
         Args:
-            images (torch.Tensor): The input images.
-            labels (torch.Tensor): The ground truth labels.
+            images (torch.Tensor | dict[str, torch.Tensor]): The input images.
+            labels (torch.Tensor | dict[str, torch.Tensor] | None, optional): The ground truth labels.
             mode (str, optional): The mode of the forward pass. Defaults to "tensor".
 
         Returns:
             torch.Tensor: The output logits or loss, depending on the training mode.
         """
-        feats = self.backbone(images)
-        if len(feats.shape) == 4 and not self.use_layer_norm_2d:  # If feats is a 4D tensor: (b, c, h, w)
-            feats = feats.view(feats.size(0), -1)  # Flatten the output of the backbone: (b, f)
-        logits = self.head(feats)
         if mode == "tensor":
-            return logits
+            return self.extract_feat(images, stage="head")
         if mode == "loss":
-            return self.loss(logits, labels)
+            feats = self.extract_feat(images, stage="neck")
+            return self.loss(feats, labels)
         if mode == "explain":
             return self._forward_explain(images)
-
+        logits = self.extract_feat(images, stage="head")
         return self.softmax(logits)
 
+    def extract_feat(
+        self,
+        inputs: dict[str, torch.Tensor] | torch.Tensor,
+        stage: str = "neck",
+    ) -> dict[str, tuple | torch.Tensor] | tuple | torch.Tensor:
+        """Extract features from the input tensor with shape (N, C, ...).
+
+        Args:
+            inputs (dict[str, torch.Tensor] | torch.Tensor): A batch of inputs. The shape of it should be
+                ``(num_samples, num_channels, *img_shape)``.
+            stage (str): Which stage to output the feature. Choose from:
+
+                - "backbone": The output of backbone network. Returns a tuple
+                  including multiple stages features.
+                - "neck": The output of neck module. Returns a tuple including
+                  multiple stages features.
+                - "pre_logits": The feature before the final classification
+                  linear layer. Usually returns a tensor.
+
+                Defaults to "neck".
+
+        Returns:
+            dict[str, tuple | torch.Tensor] | tuple | torch.Tensor: The output of specified stage.
+            The output depends on detailed implementation. In general, the
+            output of backbone and neck is a tuple.
+        """
+        if isinstance(inputs, dict):
```

Comment:
I think it should be checked if all keys for semi-sl are dict.
If not, value error can occur at L258~260.
What do you think?

Reply:
Added syntax to check for key, thanks. https://github.com/openvinotoolkit/training_extensions/pull/3614/commits/6bc9bfa48aa11b0f5b4677f2e6a489e830ceda75